File: web/frontend/src/api/skillParser.ts
/** 技能解析器（TypeScript 版）— 匹配 Python skill_parser 行为。 */

/** 用 DOM 解析器安全剥离 HTML 标签，防止正则 bypass。 */
function stripHtmlTags(html: string): string {
  if (typeof DOMParser !== "undefined") {
    const doc = new DOMParser().parseFromString(
      html.replace(/<BR\s*\/?>/gi, "\n"),
      "text/html",
    );
    return doc.body.textContent || "";
  }
  // 回退：在无 DOM 环境（测试）下用更严格的正则
  // 注意：输出仅用作 textContent，不用于 setHTML/innerHTML，XSS 风险可控
  return (
    html
      // 先处理 BR 换行
      .replace(/<BR\s*\/?>/gi, "\n")
      // 移除所有 HTML 标签（含自闭合、注释）
      .replace(/<!--[\s\S]*?-->/g, "")
      .replace(/<[^>]*>/g, "")
      // 移除 HTML 实体编码
      .replace(/&[^;]+;/g, " ")
      // 移除零宽字符和符号
      .replace(/[​-‍﻿]/g, "")
  );
}

export interface ParsedSkill {
  name: string;
  spType: string;
  trigger: string;
  effectiveMultiplier: number;
  atkBuffHint: number;
  hitCount: number;
  hasConditional: boolean;
  conditionalMult: number;
  damageType: "physical" | "magical" | "true";
  isHealing: boolean;
  spCost: number;
  initSp: number;
  duration: number;
  description: string;
  totalMult: number;
  _directAtkMult?: number;
  _equivDamageMult?: number;
}

export function parseSkill(skill: ArknightsSkill, level: number): ParsedSkill {
  const info: ParsedSkill = {
    name: skill.name ?? "",
    spType: skill.sp_type ?? "",
    trigger: skill.trigger ?? "",
    effectiveMultiplier: 1.0,
    atkBuffHint: 0,
    hitCount: 1,
    hasConditional: false,
    conditionalMult: 1.0,
    damageType: "physical",
    isHealing: false,
    spCost: 0,
    initSp: 0,
    duration: 0,
    description: "",
    totalMult: 1.0,
  };

  const levels = skill.levels ?? [];
  if (levels.length === 0) return info;

  const idx = Math.min(Math.max(level - 1, 0), levels.length - 1);
  const lv = levels[idx];
  const rawDesc = lv.description ?? "";
  const desc = stripWikiMarkup(rawDesc);
  info.description = desc;
  info.spCost = lv.sp_cost ?? 0;
  info.initSp = lv.init_sp ?? 0;
  try {
    info.duration = parseInt(String(lv.duration), 10) || 0;
  } catch {
    info.duration = 0;
  }

  // healing
  if (desc.includes("治疗") && desc.includes("攻击力")) {
    info.isHealing = true;
  }
  if (/相当于攻击力[\d.]+%的.*?(?:生命|血量)/.test(desc)) {
    info.isHealing = true;
  }

  // damage type
  if (desc.includes("真实")) info.damageType = "true";
  else if (desc.includes("法术")) info.damageType = "magical";
  else if (desc.includes("物理")) info.damageType = "physical";

  // conditional
  const condMatch = desc.match(/仅攻击到(?:一个|1个)敌人时.*?攻击力提升至(\d+(?:\.\d+)?)%/);
  if (condMatch) {
    info.hasConditional = true;
    info.conditionalMult = parseFloat(condMatch[1]) / 100;
  }

  // pure ASPD
  if (desc.includes("攻击速度") && !desc.includes("攻击力") && !desc.includes("相当于")) {
    info.effectiveMultiplier = 1.0;
    extractHitCount(info, desc);
    info.totalMult = info.effectiveMultiplier * info.hitCount;
    return info;
  }

  // ATK buff: 攻击力+XX%
  const buffMatch = desc.match(/攻击力\+(\d+(?:\.\d+)?)%/);
  if (buffMatch) {
    info.atkBuffHint = parseFloat(buffMatch[1]) / 100;
  }

  // direct ATK set: 攻击力提升至XX% (exclude conditional context)
  const beforeCond = desc.split(/仅攻击到/)[0];
  const directMatch = beforeCond.match(/攻击力提升至(\d+(?:\.\d+)?)%/);
  if (directMatch) {
    info._directAtkMult = parseFloat(directMatch[1]) / 100;
  }

  // equiv damage: 相当于攻击力XX%
  const eqMatch = desc.match(/相当于攻击力(\d+(?:\.\d+)?)%/);
  if (eqMatch) {
    info._equivDamageMult = parseFloat(eqMatch[1]) / 100;
  }

  // variant: 攻击力XX%的
  if (info._equivDamageMult === undefined) {
    const pctOfMatch = desc.match(/攻击力(\d+(?:\.\d+)?)%的/);
    if (pctOfMatch) {
      info._equivDamageMult = parseFloat(pctOfMatch[1]) / 100;
    }
  }

  extractHitCount(info, desc);
  resolveEffective(info);
  info.totalMult = info.effectiveMultiplier * info.hitCount;

  return info;
}

function resolveEffective(info: ParsedSkill) {
  const hasBuff = info.atkBuffHint > 0;
  const hasDirect = (info._directAtkMult ?? 1) > 1.0;
  const _equiv = info._equivDamageMult;
  const hasEquiv = _equiv !== undefined && (_equiv > 1.0 || (_equiv < 1.0 && _equiv !== 1.0));

  if (hasEquiv && _equiv !== undefined) {
    info.effectiveMultiplier = _equiv;
    return;
  }
  if (hasDirect && info._directAtkMult !== undefined) {
    info.effectiveMultiplier = info._directAtkMult;
    return;
  }
  if (hasBuff) {
    info.effectiveMultiplier = 1.0;
    return;
  }
  info.effectiveMultiplier = 1.0;
}

function extractHitCount(info: ParsedSkill, desc: string) {
  const hitMatch = desc.match(/(\d+)连发/) || desc.match(/(\d+)次射击/) || desc.match(/连续攻击(\d+)次/);
  if (hitMatch) {
    info.hitCount = parseInt(hitMatch[1], 10);
  }
}

function stripWikiMarkup(text: string): string {
  // 递归剥离所有 {{...}} wiki 标记（含嵌套）
  // 用深度计数器逐字符处理，确保正确处理任意深度嵌套
  let depth = 0;
  let result = "";
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (ch === '{' && text[i + 1] === '{') {
      depth++;
      i++; // 跳过第二个 {
      continue;
    }
    if (ch === '}' && text[i + 1] === '}') {
      depth--;
      i++; // 跳过第二个 }
      continue;
    }
    if (depth === 0) {
      result += ch;
    }
  }
  // 用 DOM 解析器安全剥离 HTML 标签（比正则更完整，防止 bypass）
  const cleaned = stripHtmlTags(result);
  return cleaned.replace(/\s+/g, " ").trim();
}

export function parseAutoAttack(): ParsedSkill {
  return {
    name: "普攻",
    spType: "",
    trigger: "",
    effectiveMultiplier: 1.0,
    atkBuffHint: 0,
    hitCount: 1,
    hasConditional: false,
    conditionalMult: 1.0,
    damageType: "physical",
    isHealing: false,
    spCost: 0,
    initSp: 0,
    duration: 0,
    description: "普通攻击",
    totalMult: 1.0,
  };
}

export interface ArknightsSkillLevel {
  description: string;
  sp_cost: number;
  init_sp: number;
  duration: string;
}

export interface ArknightsSkill {
  name: string;
  sp_type: string;
  trigger: string;
  levels: ArknightsSkillLevel[];
}

File: games/arknights/calc/skill_parser.py
# -*- coding: utf-8 -*-
# SPDX-License-Identifier: AGPL-3.0
"""明日方舟技能描述解析器。

从 BWIKI 爬取的技能描述 text 中提取：
- 伤害倍率 / 攻击力加成
- 连发数/段数
- 条件触发倍率
- 伤害类型 / 治疗标记
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any


@dataclass
class ParsedSkillInfo:
    """技能解析结果。

    语义说明（用于 DAG 计算）：
    - effective_multiplier: 每段伤害的最终乘数（推荐给 DAG 的 skill_multiplier）
    - atk_buff_hint: 若技能含 ATK buff，给出建议的 atk_percent_bonus
    """

    name: str = ""
    sp_type: str = ""
    trigger: str = ""

    effective_multiplier: float = 1.0
    atk_buff_hint: float = 0.0
    hit_count: int = 1
    has_conditional: bool = False
    conditional_mult: float = 1.0

    damage_type: str = "physical"
    is_healing: bool = False
    sp_cost: int = 0
    init_sp: int = 0
    duration: int = 0
    description: str = ""

    total_mult: float = 1.0

    # 原始提取值（调试用）
    _atk_buff_pct: float = 0.0
    _direct_atk_mult: float = 1.0
    _equiv_damage_mult: float = 1.0


def _strip_wiki_markup(text: str) -> str:
    text = re.sub(r"\{\{蓝色\|([^}]+)\}\}", r"\1", text)
    text = re.sub(r"\{\{color\|[^|]+\|([^}]+)\}\}", r"\1", text)
    text = re.sub(r"<BR\s*/?>", " ", text, flags=re.IGNORECASE)
    text = re.sub(r"<[^>]+>", "", text)
    """strip wiki markup。"""
    return text.strip()


def parse_skill(skill_data: dict[str, Any], level: int = 7) -> ParsedSkillInfo:
    """解析指定等级的技能参数。"""
    info = ParsedSkillInfo(
        name=skill_data.get("name", ""),
        sp_type=skill_data.get("sp_type", ""),
        trigger=skill_data.get("trigger", ""),
    )

    levels = skill_data.get("levels", [])
    if not levels:
        return info

    idx = min(max(level - 1, 0), len(levels) - 1)
    lv = levels[idx]
    raw_desc = lv.get("description", "")
    info.description = _strip_wiki_markup(raw_desc)
    info.sp_cost = lv.get("sp_cost", 0)
    info.init_sp = lv.get("init_sp", 0)

    dur_str = str(lv.get("duration", "0"))
    try:
        info.duration = int(dur_str)
    except ValueError:
        info.duration = 0

    _extract_all_params(info, raw_desc)

    return info


def _extract_all_params(info: ParsedSkillInfo, raw_desc: str) -> None:
    """全面提取所有参数（不早返回）。"""
    desc = _strip_wiki_markup(raw_desc)

    # --- 治疗检测 ---
    # 严格：只有明确说"治疗"才标记（排除"恢复技力""恢复生命上限"等）
    if "治疗" in desc and "攻击力" in desc:
        info.is_healing = True
    # "相当于攻击力XX%的生命"也视为治疗
    if re.search(r"相当于攻击力[\d.]+%的.*?(?:生命|血量)", desc):
        info.is_healing = True

    # --- 伤害类型 ---
    type_order = [("真实", "true"), ("法术", "magical"), ("物理", "physical")]
    for kw, dt in type_order:
        if kw in desc:
            info.damage_type = dt
            break

    # --- 条件触发倍率 ---
    cond = re.search(r"仅攻击到(?:一个|1个)敌人时.*?攻击力提升至(\d+(?:\.\d+)?)%", desc)
    if cond:
        info.has_conditional = True
        info.conditional_mult = float(cond.group(1)) / 100.0

    # --- 纯攻速技能 ---
    if "攻击速度" in desc and "攻击力" not in desc and "相当于" not in desc:
        info.effective_multiplier = 1.0
        _extract_hit_count(info, desc)
        _compute_total(info)
        return

    # --- 攻击力+XX% (ATK buff) ---
    buff = re.search(r"攻击力\+(\d+(?:\.\d+)?)%", desc)
    if buff:
        info._atk_buff_pct = float(buff.group(1)) / 100.0

    # --- 攻击力提升至XX% (direct ATK set) ---
    # 排除条件语句内的（如"仅攻击到一人时对其攻击力提升至150%"）
    for m in re.finditer(r"攻击力提升至(\d+(?:\.\d+)?)%", desc):
        pos = m.start()
        before = desc[max(0, pos - 15) : pos]
        if "仅攻击到" in before or "仅" in before:
            continue
        info._direct_atk_mult = float(m.group(1)) / 100.0
        break

    # --- 相当于攻击力XX% (damage conversion) ---
    eq = re.search(r"相当于攻击力(\d+(?:\.\d+)?)%", desc)
    if eq:
        info._equiv_damage_mult = float(eq.group(1)) / 100.0

    # --- 攻击力XX%的 (damage conversion, variant) ---
    if info._equiv_damage_mult == 1.0:
        pct_of = re.search(r"攻击力(\d+(?:\.\d+)?)%的", desc)
        if pct_of:
            info._equiv_damage_mult = float(pct_of.group(1)) / 100.0

    # --- 连发数 ---
    _extract_hit_count(info, desc)

    # --- 确定 effective_multiplier ---
    _resolve_effective_mult(info)

    # --- 计算总倍率 ---
    _compute_total(info)


def _resolve_effective_mult(info: ParsedSkillInfo) -> None:
    """综合各种原始值确定 effective_multiplier。"""
    has_buff = info._atk_buff_pct > 0.0
    has_direct = info._direct_atk_mult > 1.0
    has_equiv = info._equiv_damage_mult > 1.0 or (info._equiv_damage_mult < 1.0 and info._equiv_damage_mult != 1.0)

    # 规则 1: 有 equivalent damage mult → 用它
    if has_equiv:
        info.effective_multiplier = info._equiv_damage_mult
        info.atk_buff_hint = info._atk_buff_pct
        return

    # 规则 2: 有 direct ATK set → 用它
    if has_direct:
        info.effective_multiplier = info._direct_atk_mult
        info.atk_buff_hint = info._atk_buff_pct
        return

    # 规则 3: 只有 ATK buff → ATK 加成的普攻
    if has_buff:
        info.effective_multiplier = 1.0
        info.atk_buff_hint = info._atk_buff_pct
        return

    # 规则 4: 都没有 → 1.0
    info.effective_multiplier = 1.0


def _extract_hit_count(info: ParsedSkillInfo, desc: str) -> None:
    for pat in [r"(\d+)连发", r"(\d+)次射击", r"连续攻击(\d+)次"]:
        m = re.search(pat, desc)
        if m:
            info.hit_count = int(m.group(1))
            return
    """extract hit count。"""


def _compute_total(info: ParsedSkillInfo) -> None:
    info.total_mult = info.effective_multiplier * info.hit_count
    """compute total。"""


def parse_auto_attack() -> ParsedSkillInfo:
    """返回普攻的解析结果。"""
    return ParsedSkillInfo(name="普攻", description="普通攻击")
